### ISSUE TITLE
bug: Descriptors are not handled properly

### ISSUE THREAD

**Issue by pekkaklarck** [mkdocstrings/griffe#328]
If I use a [descriptor](https://docs.python.org/3/howto/descriptor.html) as at the end of this description, there are two problems with how the generated documentation for `Example.count` looks like:

1. There's no type information. It would be possible to get that from the `__get__` method.
2. The descriptor gets labels `class-atribute` and `instance-atribute` which are both somewhat wrong. Technically it is a class attribute, but so are methods and `property`s (that are also implemented as descriptors) and they don't get such a label either. Technically it's not an instance attribute, but from the usage perspective it acts like one. I believe a proper label would be `data-descriptor`, but `property` would probably be more familiar for normal users and the functionality is the same.

You can see how the example is rendered [here](https://pekkaklarck.github.io/mkdocs-experiments/0.4/api/robot/descriptor/) in a demo site that I've used for MkDocs experiments.

A variation of this is using a descriptor as a method decorator like the [setter](https://github.com/robotframework/robotframework/blob/2c995d7735765e440b123a1c064a4cfc5949d524/src/robot/utils/setter.py#L24) we [use extensively](https://github.com/robotframework/robotframework/blob/2c995d7735765e440b123a1c064a4cfc5949d524/src/robot/model/testcase.py#L70) with the [Robot Framework](https://github.com/robotframework/robotframework) project to avoid using normal propertys with dummy getters. Such attributes are currently shown as methods without any type information. You can see an example [here](https://pekkaklarck.github.io/manual/0.2/api/robot/model/testcase/#robot.model.testcase.TestCase.tags) in a site where we are creating our new Manual that will also incorporate API docs.

```python
from typing import overload, Self


class PositiveInteger[T]:

    def __set_name__(self, owner: T, name: str):
        self.name = '_' + name

    @overload
    def __get__(self, instance: None, owner: type[T]) -> Self:
        ...

    @overload
    def __get__(self, instance: T, owner: type[T]) -> int:
        ...

    def __get__(self, instance, owner) -> int | Self:
        if instance is None:
            return self
        return getattr(instance, self.name)

    def __set__(self, instance: T, value: int):
        if value <= 0:
            raise ValueError(f'{value} is not positive')
        setattr(instance, self.name, value)


class Example:
    """Example using descriptor."""

    count = PositiveInteger()
    """Count as a positive integer."""

    def __init__(self, count: int = 1):
        self.count = count
```

I'm not sure should this have been submitted to griffe instead or should I have submitted separate issues about different problems. Feel free to move or split as needed.

**Comment by pekkaklarck** [comment]
I don't think it would be unrealistic to get the correct type from the `__get__` method. It would then work with all descriptors, including `property` that _is_ a descriptor but many tools (including Griffe I assume) just handle is as a special construct. I have no idea how easy this would be, but I think it would be a good goal.

You can access descriptors both via the class and via the instance, but you get different results. In the former case you get the descriptor itself, but when you access it from the instance you get the value returned but `__get__`. Here's an example using the custom `PositiveInteger` description as well as a property and a method that are also descriptors:
```python
>>> class Example:
...     d = PositiveInteger()
...     def __init__(self):
...         self.d = 1
...     @property
...     def p(self):
...         return 2
...     def m(self):
...         return 3
...         
>>> e = Example()
>>> e.d
1
>>> e.p
2
>>> e.m
<bound method Example.m of <__main__.Example object at 0x7f2780d8a3c0>>
>>> Example.d
<PositiveInteger object at 0x7f2780e29bd0>
>>> Example.p
<property object at 0x7f2780c81a80>
>>> Example.m
<function Example.m at 0x7f2780c7d8a0>
```
If you want to learn more about descriptors, I recommend reading the excellent [Descriptor Guide](https://docs.python.org/3/howto/descriptor.html). Descriptors aren't used that widely in normal code, but Python itself uses them extensively. I'm sure their usage gets more widespread in the future, though. For example, I read somewhere that SQLAlchemy has started to used from in its models.

I would expect that our `setter` could be handled as a normal descriptor without a dedicated extension. It's easy to recognize that an attribute is a descriptor with `hasattr(attr, '__get__')` and from there it ought to be possible to handle all of them the same way. I'm interested to look at that myself, but unfortunately I don't have time for that in the near future.

**Comment by pekkaklarck** [comment]
Few more examples using the `Example` class in the above comment:
```python
>>> Example.d.__get__(e, type(e))     # Standard descriptor call requires `instance` and `type`.
1
>>> Example.d.__get__(None, type(e))  # `None` as `instance` is same as accessing from class.
<descriptor.PositiveInteger object at 0x7f2780e29bd0>
>>> Example.p.__get__(e)              # With property `type` is optional.
2
>>> Example.m.__get__(e)()            # Somewhat strange way to call a method...
3
>>> from typing import get_type_hints
>>> get_type_hints(Example.d.__get__)
{'return': typing.Union[int, typing.Self]}
```

As the end of the example demonstrates, getting the type from `__get__` is easy. There are two problems, but both ought to be relatively easy to handle:
- A properly typed `__get__` must return an union that has the descriptor itself as the other member. That type be specified at least as `typing.Self` (new in 3.11), as a forward reference like `'PositiveInteger'` and as a generic like `T`. Knowing which one of the union members is the type we are looking for can be thus hard. If it's too hard, I'd be absolutely fine with Griffe requiring the first member to be the "normal" type and latter the descriptor type.
- It's likely that there are `__get__` methods that aren't properly typed and just return the "normal" type like `int`. That's a typing bug, but I'd say Griffe could just use that type directly.

**Comment by pawamoy** [comment]
> It would then work with all descriptors, including property that is a descriptor

> I would expect that our setter could be handled as a normal descriptor without a dedicated extension. It's easy to recognize that an attribute is a descriptor with hasattr(attr, '__get__')

You seem to assume dynamic analysis here, with access to runtime objects. But by default, Griffe uses static analysis. I'm sure you understand it's another story :smile: Griffe also supports dynamic analysis though, but I'm not sure we explicitly support data descriptors (we do support method descriptors).

**Comment by pekkaklarck** [comment]
It actually occurred to be, after writing the previous comments, that this isn't as straightforward with static analysis alone. I assume methods (that are non-data descriptors) work also in static analysis, so perhaps the same approach could also be extended to data descriptors. If that's too complicated, proper data descriptor support requiring dynamic analysis ought to be fine.

**Comment by pawamoy** [comment]
I believe we can (and should) improve support for data-descriptors through both static and dynamic analysis. Data descriptors are a feature of the language, so even if it's hard, Griffe must support them. Dynamic analysis should be able to handle them directly while visiting object trees, while static analysis will probably require a built-in extension. Anyway, these are just internal details :slightly_smiling_face: 

What I note is:

- set `data-descriptor` label instead of `class-attribute` and `instance-attribute`
- try and fetch accepted type from descriptor's `__get__` method
- set `writable` label if `__set__` is implemented
- test our dynamic analysis with code that uses descriptors: if they're not properly handled, try and support them like method descriptors

After this, to support your `setter` you'll either have to rely on dynamic analysis, or write a Griffe extension as static analysis will still not be enough I believe.

**Comment by pawamoy** [comment]
Thanks for all the explanations, and for your patience :slightly_smiling_face: I've read the guide on data descriptors you shared, and I think I understand their main benefit: contrary to properties, they are easily reusable, and require much less boilerplate once implemented. Properties would have to shared via inheritance (mixins) instead of composition, and even then would be less flexible.

**Comment by pekkaklarck** [comment]
Yeah, descriptors are really handy. They look complicated first, but users typically don't need to even know them. I think they are underused but believe that will change.

Your plans sound really good. The `data-descriptor` label may look strange to some, but that's anyway probably better that more familiar but not-exactly-right `property`. The situation could be enhanced by mkdocstring by linking labels to more information similarly as type hints to external types are. I'm not sure do all labels have a suitable target at python.org, though, or should their explanations be part of mkdocstring docs.

If you implement all that, I'd expect our `setter` to work out-of-the-box at least in dynamic analysis.
